stdpopsim/models.py

```diff
@@ -137,6 +137,176 @@ def asdict(self):
+        function of time that gives inverse coalescence rate.
+
+        `num_samples_per_location` should be a list containing the
+        number of samples in each subpopulation organized by population index.
+
+        `num_steps` parameter will determine how many points along the
+        time axis are returnes.
+
+        Returns a tuple where the first value
+        is the steps (in generations) and the second is a numpy
+        ndarray that will contain the effective population sizes for
+        each subpopulation individually, for each step, as defined
+        by the model which inherits from this class.
+        """
+
+        num_pops = self.num_pops()
+        assert(len(num_samples_per_location) == num_pops)
+        P = np.zeros([num_pops**2, num_pops**2])
+        index_array = np.array(range(num_pops**2)).reshape([num_pops, num_pops])
+        for x in range(num_pops):
+            for y in range(num_pops):
+                K_x = num_samples_per_location[x]
+                K_y = num_samples_per_location[y]
+                P[index_array[x, y], index_array[x, y]] = K_x * (K_y - self._delta(x, y))
+        P = P / np.sum(P)
+        r = np.zeros(num_steps)
+        steps = np.linspace(0, end, num_steps)
+        dt = steps[1] - steps[0]
+        for j in range(num_steps):
+            N_M = self._get_N_M_at_t(steps[j])
+            pop_sizes = N_M["population_sizes"]
+            mig_matrix = N_M["migration_matrix"]
+            C = self._get_C_from_N(pop_sizes)
+            Mp = self._get_Mp_from_M(mig_matrix)
+            G = self._get_G_from_Mp_C(Mp, C)
+            P = P * np.exp(dt * G)
+            r[j] = np.sum(np.matmul(P, C)) / np.sum(P)
+        return steps, r
+
+    def num_pops(self):
+        """
+        This function returns the number populations
+        defined by the demographic model
+        """
+        ddb = msprime.DemographyDebugger(**self.asdict())
+        return len(ddb.epochs[0].populations)
+
+    # TODO Rename
+    def _get_epoch_at_t(self, t):
+        """
+        Given a time, t (in generations), find and return
+        the Epoch for which this belongs.
+        If there is no epoch that exists for the given t
+        then it returns None.
+
+        <class 'int'> -> <class 'msprime.simulations.Epoch'>
+        """
+
+        ddb = msprime.DemographyDebugger(**self.asdict())
+        epochs = ddb.epochs
+        j = 0
+        while epochs[j].end_time <= t:
+            j += 1
+        return j, epochs[j]
+
+    # TODO Rename
+    def _get_N_M_at_t(self, t):
+        """
+        Given a time, t (in generations), find and return
+        a dictionary containing:
+        1: "population_sizes" the vector N which should represent effective
+        population size for each populations at time t and,
+        2: "migration_matrix" The migration matrix for for the populations at time t
+
+        <class 'int'> -> <class 'dict'>
+        """
+
+        epochIndex, epoch = self._get_epoch_at_t(t)
```

Comment:
if you moved the `get_epoch` code into here (the only place that function is used) then you wouldn't have to create an extra DemographyDebugger...

Reply:
Agreed.